include/eeprom/core/hardware_iface.hpp

```diff
@@ -0,0 +1,73 @@
+#pragma once
+
+namespace eeprom {
+namespace hardware_iface {
+
+/**
+ * Interface to eeprom. Must be implemented in FW and Simulation
+ */
+class EEPromHardwareIface {
+  public:
+    EEPromHardwareIface() = default;
+    EEPromHardwareIface(const EEPromHardwareIface&) = default;
+    EEPromHardwareIface(EEPromHardwareIface&&) = default;
+    auto operator=(EEPromHardwareIface&&) -> EEPromHardwareIface& = default;
+    auto operator=(const EEPromHardwareIface&)
+        -> EEPromHardwareIface& = default;
+    virtual ~EEPromHardwareIface() = default;
+
+    /**
+     * Change the write protect status.
+     * @param enabled true to inhibit writing
+     */
+    virtual void set_write_protect(bool) = 0;
+};
+
+/**
+ * Class to manages the state of the write protect pin across multiple
+ * transactions.
+ *
+ * Consecutive write requests will want to have write protection disabled. The
+ * enable should happen when all the write transactions have completed.
+ *
+ * This class keeps the eeprom from having to keep track of the write protection
+ * state.
+ */
+class WriteProtector {
```

Comment:
this definitely will work as long as it's called from one task, and I don't think it'll be a problem, but this is definitely a semaphore which [freertos has an implementation of](https://www.freertos.org/xSemaphoreCreateCountingStatic.html) - maybe we should back this class with one of those?